### PR TITLE
Filter pending state manager ops while getLocalState by sequence number.

### PR DESCRIPTION
**Pull request by dannimad**
Vlad called out the fact that we should filter saved ops in the pending state manager based on their sequence number.
Interesting enough, I needed to bring back Tyler hack on one of his tests that removed savedOps, otherwise our assert ""saved op should already have a sequence number"" is triggered. If I remove the assert, we end up trying to load from a newer snapshot. 
More test coverage is needed and the feature will remain disabled until we have improve our refresh signal and test coverage.

### CODE REVIEW

File: packages/runtime/container-runtime/src/pendingStateManager.ts [hunk 1]
```diff
@@ -127,19 +128,23 @@ export class PendingStateManager implements IDisposable {
...
 		);
-		const newSavedOps = [...this.savedOps].filter(
-			(message) => message.referenceSequenceNumber >= (snapshotSequenceNumber ?? 0),
-		);
+		const newSavedOps = [...this.savedOps].filter((message) => {
+			assert(
+				message.sequenceNumber !== undefined,
+				"saved op should already have a sequence number",
+			);
+			return message.sequenceNumber >= (snapshotSequenceNumber ?? 0);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:136]: This is probably minor, but what is the semantics of snapshotSequenceNumber === undefined?
Is it just staging of the changes (and it will become non-optional soon), or it's actually a state we support?
It's not very clear what are the contract around it, i.e. is it guaranteed that snapshot sequence number will not change through rehydration process if it was undefined on save?

**Comment by vladsud** [comment]
I do not fully understand that hack that Tyler added (or the flow that caused it) - I'd need to debug that UT to learn.
But other changes look correct to me.

The fact that tests pass is great from one POV, but really bad from another POV. It basically says - we have zero coverage for these kinds of interesting cases. I'd definitely invest into some kind of either stress test (that randomly tries to hit all these cases) or a test skeleton function that has a bunch of knobs and can expose different orderings, and thus we can call it many times with all combinations of knobs to ensure system performs correctly with all such cases.

This will give us more confidence. I.e. right now, if we treat actual behavior as black box, we have two solutions that pass tests, but it's impossible for both of them to be correct (but possible for both solutions to be wrong). It would be great to have signal that more reliably tells "wrong" (it will never tell us "It's correct", but if it filters out all the cases we can come up with that are wrong, then it's enough in terms of building confidence).

File: packages/runtime/container-runtime/src/pendingStateManager.ts (hunk 1 above)
Comment (vladsud) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:136]: @dannimad, I thought a bit about Tony's point "it does not matter", and I want to share some thoughts here on that.
Looking again at existing code, I agree it should not matter, assuming the invariant about when refresh happens holds.
So maybe we are missing clear comment in the code why it's correct.

At the same time, I've shared earlier pointer to PR that shows that some ops are excluded from "dirty" calculation (it's same ID compressor ops :), but the case is broader that just that). From that angle, existing code, before or after this PR is not sufficient. As Loader layer has no clue when its safe to update snapshot. While not ideal, if you broaden the definition of "safe" to be !ConnectionManager.shouldJoinWrite(), then it would be correct, as it includes these cases.
Maybe that's it, but to be honest it's very hard to follow all these implicit contracts in the code. So, I'd rather have more explicit mechanism to ask runtime these things.

The nice thing about explicit ask is that runtime can remember updated snapshot's sequence number, and use it (or assert it's equal to snapshotSequenceNumber), and thus it's easier for readers to follow where it comes from and why it's safe assumption that all ops in pending state will / have to satisfy this requirement.

Comment (vladsud) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:136]: But I think we still have unanswered question - if it does not matter - why the hack in Tyler's test is required? It's not a question of - can test be written without this hack. It's a question - why making this change here changes the need to have this hack? Maybe it's accidental and test issue but would be great to better understand that.

Comment (dannimad) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:136]: It's a state we support. If snapshotSequenceNumber is undefined, means snapshot refresh is disabled or we didn't refresh before capturing state again, so no filtering should be applied.

Comment (dannimad) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:136]: I'll work on broadening our safe state in a subsequent PR

Comment (dannimad) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:136]: Also, I'll improve test coverage to see if we hit Tyler's issue or a test shenanigan was the cause of needing such a hack.

Comment (vladsud) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:136]: But should we always pass sequence number of a snapshot?

Comment (vladsud) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:136]: Basically, comparison with zero is wrong (or rather - not very strong). I'd rather see here a sequence number of a snapshot we will use - all the ops have to be higher than that sequence number, no matter if we use refresh or not

**Review by vladsud — APPROVED** [review]

**Comment by dannimad** [comment]
I will work in follow ups provided here in subsequent PRs and track it here https://dev.azure.com/fluidframework/internal/_workitems/edit/7891
